### ISSUE TITLE
Invalid string ("./browser.json"/>): SyntaxError: Unexpected token / in JSON

### ISSUE THREAD

**Issue by sanjeevashrivastava** [marko-js/marko#290]
Hi,
I am using Marko v3 and facing the following issue.

An error occurred while trying to compile template at path "D:\Projects\Test\Test\views\admin\Dashboard.marko". Error(s) in template: 1) [views\admin\Dashboard.marko:1:25] Invalid string ("./browser.json"/>): SyntaxError: Unexpected token / in JSON at position 16 

```
Error: An error occurred while trying to compile template at path "D:\Projects\Test\Test\views\admin\Dashboard.marko". Error(s) in template:
1) [views\admin\Dashboard.marko:1:25] Invalid string ("./browser.json"/>): SyntaxError: Unexpected token / in JSON at position 16

    at Compiler.compile (D:\Projects\Test\Test\node_modules\marko\compiler\Compiler.js:124:25)
    at Object.compileFile (D:\Projects\Test\Test\node_modules\marko\compiler\index.js:91:25)
    at loadFile (D:\Projects\Test\Test\node_modules\marko\runtime\loader.js:116:34)
    at load (D:\Projects\Test\Test\node_modules\marko\runtime\loader.js:157:16)
    at Object.load (D:\Projects\Test\Test\node_modules\marko\runtime\marko-runtime.js:296:33)
    at exports.DashboardPage (D:\Projects\Test\Test\controllers\admin\Dashboard.js:2:26)
    at Layer.handle [as handle_request] (D:\Projects\Test\Test\node_modules\express\lib\router\layer.js:82:5)
    at next (D:\Projects\Test\Test\node_modules\express\lib\router\route.js:100:13)
    at Route.dispatch (D:\Projects\Test\Test\node_modules\express\lib\router\route.js:81:3)
    at Layer.handle [as handle_request] (D:\Projects\Test\Test\node_modules\express\lib\router\layer.js:82:5)
```

My template is as below

```xml
<lasso-page package-path="./browser.json"/>
<!DOCTYPE html>
<html ng-app="MyPortal" lang="en">
  <head>
    <title>${data.title}</title>
    <meta charset="UTF-8" />
    <meta content='width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no' name='viewport'/>      
    <link href="${data.CDNUrl}/stylesheets/style.css" rel="stylesheet" type="text/css" />
    <lasso-head/>
  </head>
  <body >
    <div style="border:1px solid;border-color:#595959;width:100%;height:100%;">
      <table style="width:100%;height:100%;border:0px solid;border-collapse:collapse">
        <tr>
          <td style="height:120px;margin:0;padding:0px">
            <table style="width:100%;border:0px solid;border-collapse:collapse">
              <tr style="height:80px;background-color:#e9eaed">
                <td>
                  <a href="/">
                    <img src="${data.CDNUrl}/images/Logo.png" alt="" />
                  </a>
                </td>
              </tr>
              <tr style="height:40px;background-color:#333">
                <td>
                  <a class="link" href="/Admin/Dashboard">Dashboard</a>
                </td>
              </tr>
            </table>
          </td>
        </tr>
        <tr>
          <td style="background-color:white;margin:0;padding:0px">

          </td>
        </tr>
        <tr>
          <td style="height:50px;margin:0;padding:0px">
            
          </td>
        </tr>
      </table>
    </div>
    <lasso-body/>
    <init-widgets/>
    <browser-refresh/>
  </body>
</html>
```

Can you tell me what is the wrong with the template?

**Comment by yomed** [comment]
Hi @sanjeevashrivastava, is it possible that the JSON in your `browser.json` file is malformed?

**Comment by sanjeevashrivastava** [comment]
content of browser.json is as below

{
    "dependencies": [
        "style.css"
    ]
}

**Comment by patrick-steele-idem** [comment]
It doesn't look to be a problem in `browser.json`. The error is originally thrown by the html-js parser used by marko: https://github.com/philidem/htmljs-parser/blob/master/Parser.js#L51

@sanjeevashrivastava Are you using an older version of marko/htmljs-parser? I am not able to reproduce the problem using your exact template. Are you using an `npm-strinkwrap.json` that is locking your app to older versions of npm dependencies?

**Comment by sanjeevashrivastava** [comment]
I am using htmljs-parser@1.5.12

**Comment by sanjeevashrivastava** [comment]
I am not using npm-strinkwrap.json

**Comment by patrick-steele-idem** [comment]
That's really bizarre... That is the latest version of `htmljs-parser` and I am not able to reproduce with the exact version. Can you please attach your template file to this Github issue? I'm wondering if there are some Windows vs Unix line ending differences that are causing problems.

**Comment by yomed** [comment]
Not sure if this helps, but `at position 16` implies that the attribute name is getting parsed here.

**Comment by sanjeevashrivastava** [comment]
at position 16 / is in <lasso-page package-path="./browser.json"/> of package-path attribute

**Comment by patrick-steele-idem** [comment]
I'm still not able to reproduce. It looks like the parser is reading the string too far for you (it's going past the ending quote for the string). I see no reason for this to happen and the test I added is passing...

**Comment by sanjeevashrivastava** [comment]
i also noticed that if template is in views folder then it is working and if this template is in views/admin folder then it is throwing this error

**Comment by patrick-steele-idem** [comment]
@sanjeevashrivastava do you mind spending some more time trying to isolate the problem? Some things to try:

- Put a space after the second double quote: `<lasso-page package-path="./browser.json" />
- Reduce the template to the following:

```xml
<lasso-page package-path="./browser.json"/>
<!DOCTYPE html>
<html ng-app="MyPortal" lang="en">
  <head>
    <title>${data.title}</title>
    <meta charset="UTF-8" />
    <meta content='width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no' name='viewport'/>      
    <link href="${data.CDNUrl}/stylesheets/style.css" rel="stylesheet" type="text/css" />
    <lasso-head/>
  </head>
  <body >
    <div style="border:1px solid;border-color:#595959;width:100%;height:100%;">
      <table style="width:100%;height:100%;border:0px solid;border-collapse:collapse">
        <tr>
          <td style="height:120px;margin:0;padding:0px">
            <table style="width:100%;border:0px solid;border-collapse:collapse">
              <tr style="height:80px;background-color:#e9eaed">
                <td>
                  <a href="/">
                    <img src="${data.CDNUrl}/images/Logo.png" alt="" />
                  </a>
                </td>
              </tr>
              <tr style="height:40px;background-color:#333">
                <td>
                  <a class="link" href="/Admin/Dashboard">Dashboard</a>
                </td>
              </tr>
            </table>
          </td>
        </tr>
        <tr>
          <td style="background-color:white;margin:0;padding:0px">

          </td>
        </tr>
        <tr>
          <td style="height:50px;margin:0;padding:0px">

          </td>
        </tr>
      </table>
    </div>
    <lasso-body/>
    <init-widgets/>
    <browser-refresh/>
  </body>
</html>
```

- Do you have a different `marko.json` file in the `views/admin` folder that might be causing issue? If so, can you remove that `marko.json` to help isolate the issue?

Finally, what version of `lasso` do you have installed?

**Comment by sanjeevashrivastava** [comment]
If I put space after second " in `<lasso-page package-path="./browser.json" />` it is throwing following error

Lasso page result not found for slot "head". The <lasso-page> tag should be used to lasso the page.

Error: Lasso page result not found for slot "head". The <lasso-page> tag should be used to lasso the page.
    at render (D:\Projects\Test\Test\node_modules\lasso\taglib\slot-tag.js:51:15)
    at render (D:\Projects\Test\Test\node_modules\lasso\taglib\head-tag.js:10:5)
    at render (D:\Projects\Test\Test\views\admin\Dashboard.marko.js:20:5)
    at Object.Template.render (D:\Projects\Test\Test\node_modules\marko\runtime\marko-runtime.js:193:9)
    at exports.DashboardPage (D:\Projects\Test\Test\controllers\admin\Dashboard.js:4:14)
    at Layer.handle [as handle_request] (D:\Projects\Test\Test\node_modules\express\lib\router\layer.js:82:5)
    at next (D:\Projects\Test\Test\node_modules\express\lib\router\route.js:100:13)
    at Route.dispatch (D:\Projects\Test\Test\node_modules\express\lib\router\route.js:81:3)
    at Layer.handle [as handle_request] (D:\Projects\Test\Test\node_modules\express\lib\router\layer.js:82:5)
    at D:\Projects\Test\Test\node_modules\express\lib\router\index.js:234:24

and FYI I am using lasso@2.3.4

**Comment by sanjeevashrivastava** [comment]
[Dashboard.marko.txt](https://github.com/marko-js/marko/files/254175/Dashboard.marko.txt)
[Home.marko.txt](https://github.com/marko-js/marko/files/254176/Home.marko.txt)

if you go with above files. both files are same but Home.marko is not working while Dashboard.marko works fine.

**Comment by sanjeevashrivastava** [comment]
[Home.marko.txt](https://github.com/marko-js/marko/files/255408/Home.marko.txt)

I have tried htmljs-parser separately using the following code with attached file. it is throwing same error.

`var parser = require('htmljs-parser').createParser({
    onText: function (event) {
        // Text within an HTML element
        var value = event.value;
    },
    
    onPlaceholder: function (event) {
        //  ${<value>]} // escape = true
        // $!{<value>]} // escape = false
        var value = event.value; // String
        var escaped = event.escaped; // boolean
        var withinBody = event.withinBody; // boolean
        var withinAttribute = event.withinAttribute; // boolean
        var withinString = event.withinString; // boolean
        var withinOpenTag = event.withinOpenTag; // boolean
        var pos = event.pos; // Integer
    },
    
    onCDATA: function (event) {
        // <![CDATA[<value>]]>
        var value = event.value; // String
        var pos = event.pos; // Integer
    },
    
    onOpenTag: function (event) {
        var tagName = event.tagName; // String
        var attributes = event.attributes; // Array
        var argument = event.argument; // Object
        var pos = event.pos; // Integer
    },
    
    onCloseTag: function (event) {
        // close tag
        var tagName = event.tagName; // String
        var pos = event.pos; // Integer
    },
    
    onDocumentType: function (event) {
        // Document Type/DTD
        // <!<value>>
        // Example: <!DOCTYPE html>
        var value = event.value; // String
        var pos = event.pos; // Integer
    },
    
    onDeclaration: function (event) {
        // Declaration
        // <?<value>?>
        // Example: <?xml version="1.0" encoding="UTF-8" ?>
        var value = event.value; // String
        var pos = event.pos; // Integer
    },
    
    onComment: function (event) {
        // Text within XML comment
        var value = event.value; // String
        var pos = event.pos; // Integer
    },
    
    onScriptlet: function (event) {
        // Text within <% %>
        var value = event.value; // String
        var pos = event.pos; // Integer
    },
    
    onError: function (event) {
        // Error
        var message = event.message; // String
        var code = event.code; // String
        var pos = event.pos; // Integer
    }
});
var fs = require('fs');

fs.readFile('./views/Home.marko', 'utf8', function (err, contents) {
    parser.parse(contents);
});`

So it is confirmed that the issue is in htmljs-parser. Please try to debug it. FYI I am using Windows OS.

**Comment by patrick-steele-idem** [comment]
Thanks for the details. I was able to reproduce the issue using the uploaded file. I should be able to figure it out this morning now that I can reproduce. I'll update when I know more.

**Comment by patrick-steele-idem** [comment]
Just an update. I found the issue. The htmljs-parser is not handling a [byte order mark (BOM)](https://en.wikipedia.org/wiki/Byte_order_mark) character sequence being added by your editor. BOM characters are pretty rare, but I am working on a fix to skip over those characters.

In case you are really curious, due to the BOM characters, the parser doesn't recognize the `<` character at the beginning and it is therefore parsing in concise mode. In concise mode the parser does not look for `/>`/`>` ending character sequence.

**Comment by patrick-steele-idem** [comment]
Also, FYI:

From Wikipedia:

> The Unicode Standard permits the BOM in UTF-8, but does not require nor recommend its use. Byte order has no meaning in UTF-8.

Marko expects templates to be encoded using UTF-8. Again, we still will make the change to strip the BOM characters, but just sharing that BOM characters are not recommended in UTF-8 files.

**Comment by patrick-steele-idem** [comment]
Fixed by https://github.com/philidem/htmljs-parser/commit/baf43d342a6263bc3955a95183a15621df767cc7

New version of `htmljs-parser` published with fix: `htmljs-parser@1.5.13`

Please let me know if you see any issues. Thanks!

**Comment by sanjeevashrivastava** [comment]
Thank you very much. New version htmljs-parser@1.5.13 fixed the issue.
